### PR TITLE
Support Encoding Parquet Columns in Parallel

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Related to: https://github.com/apache/arrow-rs/issues/1718
Enables: https://github.com/apache/arrow-datafusion/pull/7655
closes https://github.com/apache/arrow-rs/pull/4871

# Rationale for this change
 

Inspired by https://github.com/apache/arrow-rs/pull/4859 but exposing a slightly different API

I have confirmed this does not appear to have any impact on benchmarks, likely because it doesn't alter any of the "hot" loops

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: parquet/src/arrow/arrow_writer/levels.rs [hunk 1]
```diff
@@ -558,10 +564,25 @@ pub(crate) struct LevelInfo {
 
     /// The maximum repetition for this leaf column
     max_rep_level: i16,
+
+    /// The arrow array
+    array: ArrayRef,
```
Comment (tustvold) [inline parquet/src/arrow/arrow_writer/levels.rs:569]: This is the major meat of the PR, by bundling the leaf array with the corresponding metadata, we get an API that is very amenable to parallelization

File: parquet/src/arrow/arrow_writer/mod.rs [hunk 2]
```diff
@@ -347,31 +349,68 @@ impl PageWriter for ArrowPageWriter {
...
+            ArrowColumnWriterImpl::ByteArray(c) => c.get_estimated_total_bytes() as _,
+            ArrowColumnWriterImpl::Column(c) => c.get_estimated_total_bytes() as _,
         }
     }
 }
 
 /// Encodes [`RecordBatch`] to a parquet row group
-pub struct ArrowRowGroupWriter {
-    writers: Vec<(SharedColumnChunk, ArrowColumnWriter)>,
+struct ArrowRowGroupWriter {
```
Comment (tustvold) [inline parquet/src/arrow/arrow_writer/mod.rs:406]: This no longer needs to be public, as this hasn't been released yet, I opted to partially revert the change in (#4850)

**Comment by tustvold** [comment]
Working on adding an example

File: parquet/src/file/writer.rs [hunk 3]
```diff
@@ -115,8 +115,7 @@ pub type OnCloseRowGroup<'a> = Box<
             Vec<Option<ColumnIndex>>,
             Vec<Option<OffsetIndex>>,
         ) -> Result<()>
-        + 'a
-        + Send,
+        + 'a,
```
Comment (tustvold) [inline parquet/src/file/writer.rs:118]: This unreleased change from #4850 is no longer necessary

**Comment by devinjdangelo** [comment]
This looks great! I will take a pass at updating https://github.com/apache/arrow-datafusion/pull/7655 to use this to test it out and report back.

**Comment by devinjdangelo** [comment]
This worked like a charm! I updated https://github.com/apache/arrow-datafusion/pull/7655 to use this branch. Your example comment was very helpful as well. I believe the datafusion PR should be handling nested columns correctly now. Performance metrics are within a margin of error vs. the previous API.

File: parquet/src/arrow/arrow_writer/mod.rs [hunk 4]
```diff
@@ -347,92 +349,213 @@ impl PageWriter for ArrowPageWriter {
...
+            }
+            ArrowColumnWriterImpl::ByteArray(c) => {
+                write_primitive(c, col.0.array().as_ref(), &col.0)?;
+            }
+        }
+        Ok(())
+    }
+
+    /// Close this column returning the [`ArrowColumnChunk`] and [`ColumnCloseResult`]
+    pub fn close(self) -> Result<(ArrowColumnChunk, ColumnCloseResult)> {
```
Comment (alamb) [inline parquet/src/arrow/arrow_writer/mod.rs:462]: While I was playing with this API, I found this combination of `(ArrowColumnChunk, ColumnCloseResult)` to be awkward to use because you always need both of them, but the api requires passing them as a pair

What would you think about wrapping them in a named struct, something like this, perhaps?

```rust
EncodedRowGroup((ArrowColumnChunk, ColumnCloseResult))
```

**Review by alamb — APPROVED** [review]
This is really neat @tustvold  and @devinjdangelo 

I had some suggestions that might make the example and the APIs a little easier to work with for mere mortals like myself 😅  -- but nothing that couldn't be done as a follow on either. 

I took a shot at using this API to make an example program that writes parquet in parallel (both row groups and column), both as a way to document how to use it / make it hopefully more discoverable as well as figuring it out myself. 

I hope to make a follow on PR with this example and some additional documentation suggestions

**Comment by alamb** [comment]
Can we also implement `Debug` for `ArrowColumnWriter`? 

```
93 | #[derive(Debug, Clone)]
   |          ----- in this derive macro expansion
...
98 |     col_writers: Vec<ArrowColumnWriter>
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `ArrowColumnWriter` cannot be formatted using `{:?}` because it doesn't implement `Debug`
   |
```

File: parquet/src/arrow/arrow_writer/mod.rs (hunk 4 above)
Comment (tustvold) [reply parquet/src/arrow/arrow_writer/mod.rs:462]: This wouldn't be usable with append_column unless you first exploded it into parts, at which point...

File: parquet/src/arrow/arrow_writer/mod.rs [hunk 5]
```diff
@@ -347,92 +349,213 @@ impl PageWriter for ArrowPageWriter {
...
+///     for c in compute_leaves(f, a).unwrap() {
+///         worker_iter.next().unwrap().1.send(c).unwrap();
+///     }
+/// }
+///
+/// // Finish up parallel column encoding
+/// for (handle, send) in workers {
+///     drop(send); // Drop send side to signal termination
+///     let (chunk, result) = handle.join().unwrap().unwrap();
+///     row_group.append_column(&chunk, result).unwrap();
```
[the remark this answers is not in the document]
Comment (tustvold) [reply parquet/src/arrow/arrow_writer/mod.rs:430]: It will just write the bytes to the Write implementation. In this case that is a Vec but if it were a File it "technically" wouldn't be a copy... FWIW this is the same as master

File: parquet/src/arrow/arrow_writer/mod.rs (hunk 4 above)
Comment (devinjdangelo) [reply parquet/src/arrow/arrow_writer/mod.rs:462]: Perhaps there could be a light wrapper around append_column that accepts `EncodedRowGroup` as @alamb defined it?

It isn't a big deal, but I agree with @alamb that it is confusing at first since the consumer of this API does not really need to concern themself with either of these two structs individually. They are more internal implementation details of arrow-rs. `EncodedRowGroup` would be more clear.

Comment (tustvold) [reply parquet/src/arrow/arrow_writer/mod.rs]: I tried to address this in https://github.com/apache/arrow-rs/pull/4871/commits/a4c17a902c11c4371eef4bd95d73a7bc8f951238 PTAL

File: parquet/src/arrow/arrow_writer/mod.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (alamb) [reply parquet/src/arrow/arrow_writer/mod.rs]: Yes, sorry - I should have been clear that I don't see anything wrong. I was just observing / validating my understanding.
